pkg/streams/utils.go

```diff
@@ -0,0 +1,185 @@
+package streams
+
+import (
+	"context"
+	"io"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// Pipes is used to present any downstream pipe, for example, containerd's cio.
+type Pipes struct {
+	Stdin io.WriteCloser
+
+	Stdout io.ReadCloser
+
+	Stderr io.ReadCloser
+}
+
+// AttachConfig is used to describe how to attach the client's stream to
+// the process's stream.
+type AttachConfig struct {
+	Terminal bool
+
+	// CloseStdin means if the stdin of client's stream is closed by the
+	// caller, the stdin of process's stream should be closed.
+	CloseStdin bool
+
+	// UseStdin/UseStdout/UseStderr can be used to check the client's stream
+	// is nil or not. It is hard to check io.Write/io.ReadCloser != nil
+	// directly, because they might be specific type, which means
+	// (typ != nil) always is true.
+	UseStdin, UseStdout, UseStderr bool
+
+	Stdin          io.ReadCloser
+	Stdout, Stderr io.Writer
+}
+
+// CopyPipes will watchs the data pipe's channel, like sticked to the pipe.
+//
+// NOTE: don't assign the specific type to the Pipes because the Std* != nil
+// always return true.
+func (s *Stream) CopyPipes(p Pipes) {
+	copyfn := func(styp string, w io.WriteCloser, r io.ReadCloser) {
+		s.Add(1)
+		go func() {
+			logrus.Debugf("start to copy %s from pipe", styp)
+			defer logrus.Debugf("stop copy %s from pipe", styp)
+
+			defer s.Done()
+			defer r.Close()
+
+			if _, err := io.Copy(w, r); err != nil {
+				logrus.WithError(err).Error("failed to copy pipe data")
+			}
+		}()
+	}
+
+	if p.Stdout != nil {
+		copyfn("stdout", s.Stdout(), p.Stdout)
+	}
+
+	if p.Stderr != nil {
+		copyfn("stderr", s.Stderr(), p.Stderr)
+	}
+
+	if s.stdin != nil && p.Stdin != nil {
+		go func() {
```

Comment:
Not controlled by waitGroup. Because it may be suddenly closed by user. Just comment to record this.